mount/titus-mount-bind.c

```diff
@@ -0,0 +1,98 @@
+#define _GNU_SOURCE
+#include <errno.h> // for errno
+#include <stdio.h> // for perror, fprintf, snprintf, stderr
+#include <stdlib.h> // for getenv, exit
+/* setns */
+#include <assert.h> // for assert
+#include <fcntl.h> // for open, openat, AT_FDCWD, AT_RECURSIVE, O_PATH
+#include <limits.h> // for PATH_MAX
+#include <linux/mount.h> // for MOVE_MOUNT_F_EMPTY_PATH, OPEN_TREE_CLOEXEC
+#include <sched.h> // for setns, CLONE_NEWNS
+#include <syscall.h> // for __NR_fsopen, __NR_move_mount, __NR_open_tree
+#include <unistd.h> // for syscall, pid_t
+
+// Only to make it easier to build on older kernels
+#  define AT_RECURSIVE		0x8000
+
+#define E(x)                                                                   \
+	do {                                                                   \
+		if ((x) == -1) {                                               \
+			perror(#x);                                            \
+			exit(1);                                               \
+		}                                                              \
+	} while (0)
+
+static inline int pidfd_open(pid_t pid, unsigned int flags)
+{
+	return syscall(__NR_pidfd_open, pid, flags);
+}
+
+static inline int fsopen(const char *fs_name, unsigned int flags)
+{
+	return syscall(__NR_fsopen, fs_name, flags);
+}
+
+static inline int move_mount(int from_dfd, const char *from_pathname,
+			     int to_dfd, const char *to_pathname,
+			     unsigned int flags)
+{
+	return syscall(__NR_move_mount, from_dfd, from_pathname, to_dfd,
+		       to_pathname, flags);
+}
+
+static inline int open_tree(int dfd, const char *filename, unsigned int flags)
+{
+	return syscall(__NR_open_tree, dfd, filename, flags);
+}
+
+static void switch_into_mount_namespace(int nsfd)
+{
+	int ret;
+	int mnt_fd = openat(nsfd, "mnt", O_RDONLY);
+	assert(mnt_fd != -1);
+	ret = setns(mnt_fd, CLONE_NEWNS);
+	if (ret == -1) {
+		perror("setns mnt");
+		exit(1);
+	}
+}
+
+int main(int argc, char *argv[])
+{
+	int nsfd, mfd;
+	/*
+	 * We do this because parsing args is a bigger pain than passing
+	 * via environment variable, although passing via environment
+	 * variable has a "cost" in that they are limited in size
+	 */
+	const char *source = getenv("MOUNT_SOURCE");
+	char *target = getenv("MOUNT_TARGET");
+	const char *pid1dir = getenv("TITUS_PID_1_DIR");
+
+	errno = 0;
+	assert(errno == 0);
+
+	if (!(target && source && pid1dir)) {
+		fprintf(stderr,
+			"Usage: must provide MOUNT_SOURCE, MOUNT_TARGET, and TITUS_PID_1_DIR env vars");
+		return 1;
+	}
+
+	/* This nsfd is used to extract other namespaces to switch to, similar to a pidfd */
+	char pid1_ns_path[PATH_MAX];
```

Comment:
*Ideally* keep all variable declarations at the top (even though that's a C89 thing)